minetest/minetest
# Detect air-only blocks instead of day/night differences

We have an optimization that avoids sending blocks to the client when day and night lighting does not differ. Above ground this avoids a lot of blocks being sent to the client.

The problem is that (a) It does not work anywhere where no day-light reaches, and (b) solid, non-transparent blocks at the surface should be sent, but are currently not (since they are solid their day/night is the same).

The main benefit of this optimization is avoiding air-blocks being sent. (It literally avoid 10000's of blocks being sent - with or without this PR.)

So this PR makes this explicit:

It detects air-only blocks, and never sends them to the client if they are not very close - as before.

If you look at the **old code** the logic is basically: If day/night lighting is identical, or the block is all air, then optimize if the block is not close and above ground.

Now the **new logic** is: If the block is all air then optimize if the block is not close.

- Goal of the PR
Remove visual glitches, allow optimizing below ground block-sending, simplify.

- How does the PR work?
See above

- Does it resolve any reported issue?
Not sure. I think we've have reports of occasionally visibly missing surface blocks.

- Does this relate to a goal in [the roadmap](https://github.com/minetest/minetest/blob/master/doc/direction.md)?
No idea.

- If not a bug fix, why is this PR needed? What usecases does it solve?
See above

## To do

This PR is Ready for Review.

## How to test

- Join an existing world. Above ground, check for any missing blocks (there should not be any, there have been before if there were all solid blocks visible at the surface.)
- Same below ground. (existing world)
- Do the same with a new world.
- In all cases you can also check the number of loaded blocks (F6), that number should not change significantly above ground, and it should significantly decrease below ground, especially in large caverns.
- Also test opening a world with a server with this PR, and then re-open this world with a server without this PR. Everything should still work as expected. (You might even notice that some visual glitches from before are now gone even with an old server, because that day/night-diff bit is now fixed to mean !isAir)
